src/main/scala/temple/ast/Templefile.scala

```diff
@@ -9,17 +10,42 @@ case class Templefile(
   projectName: String,
   projectBlock: ProjectBlock = ProjectBlock(),
   targets: Map[String, TargetBlock] = Map(),
-  services: Map[String, ServiceBlock] = Map(),
+  var services: Map[String, GeneratedBlock] = Map(),
 ) extends TempleNode {
   // Inform every child node of their parent, so that they can access the project information
   for (block <- Iterator(projectBlock) ++ targets.valuesIterator ++ services.valuesIterator) {
     block.setParent(this)
   }
 
-  def servicesWithPorts: Iterable[(String, ServiceBlock, Ports)] =
+  /**
+    * The services provided in the templefile, i.e the ServiceBlocks, not the AuthServiceBlocks
+    * @return
+    */
+  def providedServices: Map[String, ServiceBlock] = services.collect {
+    case (name: String, service: ServiceBlock) => name -> service
+  }
+
+  def allServicesWithPorts: Iterable[(String, GeneratedBlock, Ports)] =
     services
-      .zip(Iterator.from(1025, step = 2)) // 1024 is reserved for auth service
-      .map { case ((name, service), port) => (name, service, Ports(port, port + 1)) }
+      .zip(Iterator.from(ProjectConfig.serviceStartPort, step = 2)) // 1024 is reserved for auth service
+      .map {
+        case ((name, service), port) =>
+          if (service == AuthServiceBlock) (name, service, Ports(ProjectConfig.authPort, ProjectConfig.authMetricPort))
+          else (name, service, Ports(port, port + 1))
+      }
+
+  def providedServicesWithPorts: Iterable[(String, ServiceBlock, Ports)] =
+    allServicesWithPorts.collect { case (name, service: ServiceBlock, ports) => (name, service, ports) }
+
+  /**
+    * Add an extra service to an exisiting Templefile - useful for adding auth when needed.
+    * @param name - the name of the new service
+    * @param block - the [[ServiceBlock]] to add
+    */
+  def addService(name: String, block: GeneratedBlock): Unit = {
+    block.setParent(this)
+    services += name -> block
```

Comment:
and if we’re setting the parent of AuthServiceBlock, then we’ll need to change AuthServiceBlock to a (case) class rather than a case object, otherwise that parent will persist (including across test runs etc).

Reply:
One solution (and stops a port being "wasted" on auth):

Keep a constructor argument `providedServices: Map[String, ServiceBlock] = Map()`, then use these as attributes to Templefile

```scala
// Whether or not to generate an auth service - based on whether any service has #auth
private val usesAuth = templefile.services.exists {
  case (_, service) => service.lookupMetadata[ServiceAuth].nonEmpty
}

val providedServicesWithPorts: Iterable[(String, ServiceBlock, Ports)] =
  providedServices
    .zip(Iterator.from(ProjectConfig.serviceStartPort, step = 2))
    .map { case ((name, service), ports) => (name, service, ports) }

val allServicesWithPorts: Iterable[(String, GeneratedBlock, Int)] = {
  if (usesAuth) {
    val authBlock = AuthServiceBlock()
    authBlock.setParent(this)
    Iterable((auth, authBlock, Ports(ProjectConfig.authPort, ProjectConfig.authMetricPort))) ++ providedServicesWithPorts
  } else providedServicesWithPorts

val allServices: Map[String, GeneratedBlock] = allServicesWithPorts.map { case (name, service, ports) => name, service) }.toMap
```